File: core/rospy/src/rospy/service.py
# Software License Agreement (BSD License)
#
# Copyright (c) 2008, Willow Garage, Inc.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#
#  * Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
#  * Redistributions in binary form must reproduce the above
#    copyright notice, this list of conditions and the following
#    disclaimer in the documentation and/or other materials provided
#    with the distribution.
#  * Neither the name of Willow Garage, Inc. nor the names of its
#    contributors may be used to endorse or promote products derived
#    from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.
#
# Revision $Id$

import logging
import traceback

import roslib.scriptutil

from rospy.core import *
from rospy.registration import set_service_manager, Registration, get_registration_listeners
from rospy.transport import *

logger = logging.getLogger('rospy.service')

## alias of roslib.message.ServiceDefinition
import roslib.message
ServiceDefinition = roslib.message.ServiceDefinition

## \ingroup clientapi
## Exception class for service errors
class ServiceException(Exception): pass

## superclass for storing service information
class _Service(object):
    def __init__(self, name, service_class):
        self.name = resolve_name(name) #services remap as well
        self.service_class = service_class
        self.request_class = service_class._request_class
        self.response_class = service_class._response_class
        self.uri = None #initialize attr

## Keeps track of currently registered services in the ROS system
class ServiceManager(object):
    def __init__(self):
        self.map = {} # {name : Service}
        self.lock = threading.RLock()

    ## @param self
    ## @return [(str, str)]: List of (service_name, service_uri)  for all registered services.
    def get_services(self):
        try:
            self.lock.acquire()
            ret_val = []
            for name, service in self.map.iteritems():
                ret_val.append((name, service.uri))
            services = self.map.values()
        finally:
            self.lock.release()
        return ret_val
    
    ## Unregister all registered services
    ## @param self 
    def unregister_all(self):
        self.map.clear()
    
    ## Register service with ServiceManager and ROS master
    ## @param self
    ## @param service_name str: name of service (resolved)
    ## @param service Service        
    def register(self, service_name, service):
        err = None
        try:
            self.lock.acquire()
            if service_name in self.map:
                err = "service [%s] already registered"%service_name
            else:
                self.map[service_name] = service
                
            # NOTE: this call can potentially take a long time under lock and thus needs to be reimplmented
            get_registration_listeners().notify_added(service_name, service.uri, Registration.SRV)
        finally:
            self.lock.release()

        if err:
            raise ServiceException(err)
        
    ## Unregister service with ServiceManager and ROS Master
    ## @param self
    ## @param service_name str: name of service
    ## @param service Service: service implementation
    def unregister(self, service_name, service):
        try:
            self.lock.acquire()
            curr = self.map.get(service_name, None)
            if curr == service:
                del self.map[service_name]
                
            # NOTE: this call can potentially take a long time under lock
            get_registration_listeners().notify_removed(service_name, service.uri, Registration.SRV)                
        finally:
            self.lock.release()

    ## @param self
    ## @param service_name str: name of service
    ## @return Service: service implementation
    def get_service(self, service_name):
        return self.map.get(service_name, None)

set_service_manager(ServiceManager())

File: cturtle/core/roscpp/src/libros/service.cpp
/*
 * Copyright (C) 2008, Morgan Quigley and Willow Garage, Inc.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *   * Redistributions of source code must retain the above copyright notice,
 *     this list of conditions and the following disclaimer.
 *   * Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in the
 *     documentation and/or other materials provided with the distribution.
 *   * Neither the names of Stanford University or Willow Garage, Inc. nor the names of its
 *     contributors may be used to endorse or promote products derived from
 *     this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include "ros/service.h"
#include "ros/connection.h"
#include "ros/service_server_link.h"
#include "ros/service_manager.h"
#include "ros/transport/transport_tcp.h"
#include "ros/poll_manager.h"
#include "ros/init.h"
#include "ros/names.h"
#include "ros/this_node.h"
#include "ros/header.h"

using namespace ros;

bool service::exists(const std::string& service_name, bool print_failure_reason)
{
  std::string mapped_name = names::resolve(service_name);

  std::string host;
  uint32_t port;

  if (ServiceManager::instance()->lookupService(mapped_name, host, port))
  {
    TransportTCPPtr transport(new TransportTCP(0, TransportTCP::SYNCHRONOUS));

    if (transport->connect(host, port))
    {
      M_string m;
      m["probe"] = "1";
      m["md5sum"] = "*";
      m["callerid"] = this_node::getName();
      m["service"] = mapped_name;
      boost::shared_array<uint8_t> buffer;
      uint32_t size = 0;;
      Header::write(m, buffer, size);
      transport->write((uint8_t*)&size, sizeof(size));
      transport->write(buffer.get(), size);
      transport->close();

      return true;
    }
    else
    {
      if (print_failure_reason)
      {
        ROS_INFO("waitForService: Service [%s] could not connect to host [%s:%d], waiting...", mapped_name.c_str(), host.c_str(), port);
      }
    }
  }
  else
  {
    if (print_failure_reason)
    {
      ROS_INFO("waitForService: Service [%s] has not been advertised, waiting...", mapped_name.c_str());
    }
  }

  return false;
}

bool service::waitForService(const std::string& service_name, ros::Duration timeout)
{
  std::string mapped_name = names::resolve(service_name);

  Time start_time = Time::now();

  bool printed = false;
  bool result = false;
  while (ros::ok())
  {
    if (exists(service_name, !printed))
    {
      result = true;
      break;
    }
    else
    {
      printed = true;

      if (timeout >= Duration(0))
      {
        Time current_time = Time::now();

        if ((current_time - start_time) >= timeout)
        {
          return false;
        }
      }

      Duration(0.02).sleep();
    }
  }

  if (printed && ros::ok())
  {
    ROS_INFO("waitForService: Service [%s] is now available.", mapped_name.c_str());
  }

  return result;
}

bool service::waitForService(const std::string& service_name, int32_t timeout)
{
  return waitForService(service_name, ros::Duration(timeout / 1000.0));
}
